code/week7/main.cpp

```diff
@@ -0,0 +1,206 @@
+#include <iostream>
+#include <set>
+#include <unordered_set>
+#include <map>
+#include <unordered_map>
+#include <list>
+#include <stack>
+#include <queue>
+#include <memory>
+
+#include "generic_class.h"
+
+// TODO clean this method up
+template<class T>
+void print_data(T iterator_begin, T iterator_end, std::string name) {
+    std::cout << "Printing " << name << std::endl;
+    for(;iterator_begin != iterator_end; iterator_begin++) {
+        std::cout << *iterator_begin << ", ";
+    }
+    std::cout << "\n\n";
+}
+/* with set<int> it would convert to something like this
+void print_data(set<int>::iterator iterator_begin, set<int>::iterator iterator_end, std::string name) {
+    std::cout << "Printing " << name << std::endl;
+    for(;iterator_begin != iterator_end; iterator_begin++) {
+        std::cout << *iterator_begin << ", ";
+    }
+    std::cout << "\n\n";
+}
+ */
+
+template <class Foo>
+Foo getMax(Foo a, Foo b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+/* when getMax<int> the method would look like
+int getMax(int a, int b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+*/
+/*
+when getMax<double> the method would look like
+double getMax(double a, double b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+*/
+
+// you can have multiple templates in a function
+template <class Foo, class Bar>
+Bar getMin(Foo a, Bar b) {
```

Comment:
This is a little bit... weird. Since we're choosing `Bar` as our return type, we can have a possible loss of precision when using this function (and maybe some casting problems, when using it with non-trivial types).

For example, this patch:

```
From e5fdaf5e8eb5968adc30588f3237bb914b59d24e Mon Sep 17 00:00:00 2001
From: Jay Kamat <jaygkamat@gmail.com>
Date: Sat, 19 Aug 2017 19:24:47 -0400
Subject: [PATCH] Example for breaking return type templates

---
 code/week7/main.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/code/week7/main.cpp b/code/week7/main.cpp
index 5fbef94..f69e7e6 100644
--- a/code/week7/main.cpp
+++ b/code/week7/main.cpp
@@ -59,7 +59,7 @@ double getMax(double a, double b) {
 
 // you can have multiple templates in a function
 template <class Foo, class Bar>
-Bar getMin(Foo a, Bar b) {
+Foo getMin(Foo a, Bar b) {
     if(a < b) {
         return a;
     } else {
@@ -194,7 +194,7 @@ int main() {
     cout << "\n\n";
     cout << "the max of 7,8 is = " << getMax<int>(7, 8) << endl;
     cout << "the max of 7.0, 8.0 is = " << getMax<double>(7.0, 8.0) << endl;
-    cout << "the min of 7, 8.0 is = " << getMin<int, double>(7, 8.0) << endl;
+    cout << "the min of 7.1, 8.0 is = " << getMin<int, double>(8, 7.1) << endl;
 
     cout << "\n\n";
 
-- 
2.11.0
```

prints out `the min of 7.1, 8.0 is = 7`, which looses the `.1` that was inputed. I don't know of a good way of working around it while keeping the example. If you  have any good ideas, feel free to use them, but I don't have any.